### DISCUSSION TITLE
"Mirrored" hinge constraint

### DISCUSSION THREAD

**Discussion by mihe** [jrouwe/JoltPhysics#374] [Q&A] [answered]
First of all, happy holidays and a happy new year! :)

I'm in the process of implementing the hinge constraint in my Godot extension and it turned out to be more of a head-scratcher than I expected, so I was curious to hear your thoughts on it.

It seems to me like Godot and Jolt have different kinds of hinge constraints. Jolt has what I thought would be universal interpretation of a hinge constraint, where you constrain (what Jolt calls) the normal axes of the two bodies to always be within an angular span on some plane, dictated by the min/max limits.

Godot, on the other hand, seems to have what I can best describe as a "mirrored" hinge constraint, which seems to be a carry-over from Bullet from what I can tell. This type of hinge seems to constrain each body to its own angular span, both of which are dictated by a shared set of min/max limits. I guess in theory this could let you have one body with full unhindered movement around the hinge whereas the other one is constrained, but since they both share the same min/max limits this isn't possible in practice.

Hopefully the diagram below illustrates the difference, in case my explanation isn't clear.

I guess my question is: am I understanding Jolt's hinge constraint correctly, and if so, is there some way to emulate the behavior of Godot/Bullet either through some permutation of `HingeConstraintSettings` or perhaps through multiple `SixDOFConstraint` or some other type of constraint?

![Diagram illustrating the difference](https://user-images.githubusercontent.com/4884246/209543670-4ded610d-e5b9-4d93-a1de-699063b8674a.png)

**Comment by jrouwe** [comment, marked as answer]
Hello, happy holidays to you too!

I've downloaded godot (both source and binary) to see how it works there. I've created a very simple scene with 2 RigidBody3D objects and a HingeJoint3D and it seems to work as I would expect it to work:

![image](https://user-images.githubusercontent.com/1621693/209714443-1a01fab1-164d-458e-ac57-3166d4af436e.png)

I can tick the box to enable the limits and then I only see 1 min and 1 max value. The hinge seems to rotate around Z of the 'Transform' property and the pivot is specified by the position of the 'Transform' property.

Looking at the code, there seem to be 2 functions that need to be implemented:

```void GodotPhysicsServer3D::joint_make_hinge(RID p_joint, RID p_body_A, const Transform3D &p_frame_A, RID p_body_B, const Transform3D &p_frame_B)```

and

```GodotPhysicsServer3D::joint_make_hinge_simple(RID p_joint, RID p_body_A, const Vector3 &p_pivot_A, const Vector3 &p_axis_A, RID p_body_B, const Vector3 &p_pivot_B, const Vector3 &p_axis_B)```

The first function supplies the full reference frame for the hinge: p_frame_A and p_frame_B. I think to convert this to a Jolt hinge you would have to do something like (bit of pseudo code here, hope you understand):

```
HingeConstraintSettings s;
s.mSpace = EConstraintSpace::LocalToBodyCOM; // The reference frames supplied by godot seem to be local to body A and B
s.mPoint1 = p_frame_A.origin - GetJoltShape(p_body_A)->GetCenterOfMass(); // Jolt reasons from COM, godot probably from body local space
s.mHingeAxis1 = p_frame_A.basis.get_column(2); // Looks like axis Z is used for hinge axis
s.mNormalAxis1 = p_frame_A.basis.get_column(0); // And it looks like axis X is used as the axis that defines the 'zero' angle
s.mPoint2 = p_frame_B.origin - GetJoltShape(p_body_B)->GetCenterOfMass();
s.mHingeAxis2 = p_frame_B.basis.get_column(2);
s.mNormalAxis2 = p_frame_B.basis.get_column(0);
```

For the 2nd one you only get the pivots and hinges so you have to come up with a 'normal axis' yourself. The rest would work something like:

```
HingeConstraintSettings s;
s.mSpace = EConstraintSpace::LocalToBodyCOM; // The reference frames supplied by godot seem to be local to body A and B
s.mPoint1 = p_pivotA - GetJoltShape(p_body_A)->GetCenterOfMass(); // Jolt reasons from COM, godot probably from body local space
s.mHingeAxis1 = p_axis_A;
s.mNormalAxis1 = ... // this one needs to be calculated similar to how its done in 2nd GodotHingeJoint3D::GodotHingeJoint3D constructor
s.mPoint2 = p_pivot_B - GetJoltShape(p_body_B)->GetCenterOfMass();
s.mHingeAxis2 = p_axis_B;
s.mNormalAxis2 = ... // See hint at mNormalAxis1
```

The actual limits seem to be supplied post construction, so you need to implement something like ```GodotHingeJoint3D::set_param``` and catch it setting HINGE_JOINT_LIMIT_LOWER and HINGE_JOINT_LIMIT_UPPER and then call Jolt's HingeConstraint::SetLimits.

So I don't see what you mean with the 2nd picture, as far as I can see it works exactly as in Jolt. Bullet seems to work the same too. If I look at btHingeConstraint there's the same 2 reference frames being supplied and only 1 set of limits.

**Reply by mihe** [reply]
(I just realized I had mistakenly posted my reply as a new answer, so I moved it here. Forgive the notification spam.)

Thank you for taking such an in-depth look!

You are correct that there's only one set of min/max limits for Jolt, Godot Physics and Bullet alike, but at least for Godot Physics (and presumably Bullet) that set of limits is applied individually for each of the two bodies.

I've set up a small test scene similar to what my diagram showed, with a rigid body on either side of the hinge. Godot's own visualizer/gizmo for the hinge joint might illustrate the behavior I'm trying to get across:

https://user-images.githubusercontent.com/4884246/209726591-7d742960-3d88-4633-a9e4-6e3151d599a5.mp4

As you can hopefully see, the hinge is visualized as having two "pie slices", one facing each object and both sharing the same upper/lower limit, but applied individually.

If I then set the upper and lower limits to +30 and -30 respectively, and write a small script that interchangeably "freezes" one body (makes it static) and applies a force to the other body, it then looks like this at runtime:

https://user-images.githubusercontent.com/4884246/209726600-ae773834-6aee-4e13-84fd-93e61e5344e7.mp4

I would assume, if the limits were applied like they seem to be in Jolt, that the two bodies would share a single "pie slice", meaning they would be constrained to both be within the same single 60-degree span. Instead they seem to be constrained to their own individual 60-degree span, which reflects what the editor visualizer/gizmo shows as well.

It's not impossible that I've completely misinterpreted Jolt's behavior and that it does in fact line up with what the above videos show, but from my (admittedly somewhat crude) testing it seems to differ from Godot with regards to how the limits are applied.

Here's the Godot project if you'd like to take a look. It was made in the latest snapshot, [4.0 Beta 10](https://godotengine.org/en/article/dev-snapshot-godot-4-0-beta-10), but I would assume `master` works as well: [Godot_Hinge.zip](https://github.com/jrouwe/JoltPhysics/files/10310960/Godot_Hinge.zip)

**Reply by jrouwe** [reply]
I agree that this is really weird, I played around with your scene a little bit:

[Godot_Hinge.zip](https://github.com/jrouwe/JoltPhysics/files/10313408/Godot_Hinge.zip)

What I did is:
- Change the spheres to cubes (the spheres roll, that makes things really confusing from a 2D perspective)
- Make the left body 1000 kg so that it doesn't move along as easily
- Changed the hinge to limit between 0 and 45 degrees
- Changed the script to only apply 1 impulse to the right body

If you run this you can see that the constraint doesn't behave in the way that the drawing of the constraint suggests. It first bends to the 45 degrees limit and then bounces back to the 0 degrees limit (and not to -45 degrees like the drawing of the constraint suggests in the editor). This is exactly the way I would expect things to work (and is how Jolt would work).

I'm guessing that when you freeze / unfreeze the bodies that the hinge gets reinitialized using the current world space orientations of the bodies, so that it appears that the hinge limits are bigger than they actually are.

I have to say that the math in GodotHingeJoint3D::setup and GodotHingeJoint3D::solve looks pretty familiar to me and I'm almost 100% sure that it is the same math as the Jolt hinge constraint uses.

**Reply by mihe** [reply]
> If you run this you can see that the constraint doesn't behave in the way that the drawing of the constraint suggests.

After making a more elaborate test without the freezing I've come to that same realization.

It was certainly unfortunate that the freezing somewhat reinforced my skewed theory. I could have avoided a lot of head-scratching if I had just set up a proper test scene with actual dynamic bodies and better controls. Lesson learned. :)

I'll take another crack at implementing the hinges with these new tests and the pseudocode you provided.

Thank you again for taking such an in-depth look at my problem. I certainly didn't expect you to fire up Godot, much less dig through their source code.

**Reply by jrouwe** [reply]
No problem. I'm very interested in having a godot physics extension, so if there's anything else I can help out with let me know!

**Reply by mihe** [reply]
It took me a couple of days to get back to this, but I managed to implement hinges with motors and everything, in part thanks to the pseudocode you provided. You can see them in action in the clip below.

https://user-images.githubusercontent.com/4884246/210648500-ab5ef786-ffd4-4343-b9f0-4ee154a4dd9f.mp4

Next up is implementing a more sophisticated implementation of `JPH::DebugRenderer`, then I'll tackle the rest of the joint types after that.

**Reply by jrouwe** [reply]
Nice!
